staging/src/k8s.io/api/resource/v1alpha2/namedresources.go

```diff
@@ -0,0 +1,119 @@
+type NamedResourcesAttribute struct {
+	// Name is unique identifier among all resource instances managed by
+	// the driver on the node. It must be a DNS subdomain.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	NamedResourcesAttributeValue `json:",inline" protobuf:"bytes,2,opt,name=attributeValue"`
+}
+
+// The Go field names below have a Value suffix to avoid a conflict between the
+// field "String" and the corresponding method. That method is required.
+// The Kubernetes API is defined without that suffix to keep it more natural.
+
+// NamedResourcesAttributeValue must have one and only one field set.
+type NamedResourcesAttributeValue struct {
+	// QuantityValue is a quantity.
+	QuantityValue *resource.Quantity `json:"quantity,omitempty" protobuf:"bytes,6,opt,name=quantity"`
+	// BoolValue is a true/false value.
+	BoolValue *bool `json:"bool,omitempty" protobuf:"bytes,2,opt,name=bool"`
+	// IntValue is a 64-bit integer.
+	IntValue *int64 `json:"int,omitempty" protobuf:"varint,7,opt,name=int"`
+	// IntSliceValue is an array of 64-bit integers.
+	IntSliceValue *NamedResourcesIntSlice `json:"intSlice,omitempty" protobuf:"varint,8,rep,name=intSlice"`
+	// StringValue is a string.
+	StringValue *string `json:"string,omitempty" protobuf:"bytes,5,opt,name=string"`
+	// StringSliceValue is an array of strings.
+	StringSliceValue *NamedResourcesStringSlice `json:"stringSlice,omitempty" protobuf:"bytes,9,rep,name=stringSlice"`
+	// VersionValue is a semantic version according to semver.org spec 2.0.0.
+	VersionValue *string `json:"version,omitempty" protobuf:"bytes,10,opt,name=version"`
+}
+
+// NamedResourcesIntSlice contains a slice of 64-bit integers.
+type NamedResourcesIntSlice struct {
+	// Ints is the slice of 64-bit integers.
+	//
+	// +listType=atomic
+	Ints []int64 `json:"ints" protobuf:"bytes,1,opt,name=ints"`
+}
+
+// NamedResourcesStringSlice contains a slice of strings.
+type NamedResourcesStringSlice struct {
+	// Strings is the slice of strings.
+	//
+	// +listType=atomic
+	Strings []string `json:"strings" protobuf:"bytes,1,opt,name=strings"`
+}
+
+// NamedResourcesRequest is used in ResourceRequestModel.
+type NamedResourcesRequest struct {
+	// Selector is a CEL expression which must evaluate to true if a
+	// resource instance is suitable. The language is as defined in
+	// https://kubernetes.io/docs/reference/using-api/cel/
+	//
+	// In addition, for each type NamedResourcesin AttributeValue there is a map that
+	// resolves to the corresponding value of the instance under evaluation.
+	// For example:
+	//
+	//    attributes.quantity["a"].isGreaterThan(quantity("0")) &&
+	//    attributes.stringslice["b"].isSorted()
+	Selector string `json:"selector" protobuf:"bytes,1,name=selector"`
+}
+
+// NamedResourcesFilter is used in ResourceFilterModel.
+type NamedResourcesFilter struct {
+	// Selector is a selector like the one in Request. It must be true for
```

Comment:
Just repeat the documentation, instead of doing this cross-reference.
Unless you cannot set one without setting the other?

But why not just one type for both NamedResourcesRequest and NamedResourcersFilter?
They have the same intent, just at different levels.

Reply:
> Just repeat the documentation, instead of doing this cross-reference.

Okay.

> But why not just one type for both NamedResourcesRequest and NamedResourcersFilter?

We've been bitten in the past when reusing the same struct in different contexts (see ResourceRequirements vs. what is now VolumeResourceRequirements) when extending that struct later.

I can imagine that we might want to add fields to one, but not the other struct in the future. Let's prepare for that by keeping the structs separate, even if they currently contain the same field.